### PR TITLE
KNOX-2253 - Add support to handle Livy server redirects

### PR DESCRIPTION
**Pull request by RogPodge**
## What changes were proposed in this pull request?

Currently knox gateway does not seemlessly handle Livy redirect responses like it does with Yarn resource manager. We need a patch to knox gateway to handle these redirect responses on server side, and makes some adjustments to the service definition to address issues like KNOX-1609

## How was this patch tested?

This patch was tested on a local deployment of Knox gateway against a Livy Instance that was using the following HA patch: https://github.com/RogPodge/incubator-livy/tree/LivyHARedirect

### CODE REVIEW

File: gateway-service-definitions/src/main/resources/services/livy/0.4.0/rewrite.xml [hunk 1]
```diff
@@ -16,33 +16,10 @@
    limitations under the License.
```
Comment (risdenk) [inline gateway-service-definitions/src/main/resources/services/livy/0.4.0/rewrite.xml]: Don't make these unrelated changes? You look to be trying to remove `/v1` from the url handling which isn't necessary.

File: gateway-service-definitions/src/main/resources/services/livy/0.4.0/service.xml [hunk 2]
```diff
@@ -17,9 +17,9 @@
 -->
 <service role="LIVYSERVER" name="livy" version="0.4.0">
   <routes>
-    <route path="/livy/**?**"/>
-    <route path="/livy"/>
-    <route path="/livy/"/>
+    <route path="/livy/v1/**?**"/>
+    <route path="/livy/v1"/>
+    <route path="/livy/v1/"/>
```
Comment (risdenk) [inline gateway-service-definitions/src/main/resources/services/livy/0.4.0/service.xml]: Don't make these changes.

File: gateway-service-livy/pom.xml [hunk 3]
```diff
@@ -58,5 +65,10 @@
             <artifactId>gateway-test-utils</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (risdenk) [inline gateway-service-livy/pom.xml]: This shouldn't be necessary - Should be handled automatically for all submodules.

File: gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaBaseDispatcher.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.atomic.AtomicInteger;
+
+class  LivyHaBaseDispatcher extends DefaultDispatch {
```
Comment (risdenk) [inline gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaBaseDispatcher.java]: This whole class at first glance looks like a straight copy of `DefaultHaDispatch`? So why not just extend `DefaultHaDispatch`?

File: gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaBaseDispatcher.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    private static final String FAILOVER_COUNTER_ATTRIBUTE = "dispatch.ha.failover.counter";
+    private static final String LOCATION = "Location";
+    private static final RMMessages LOG = MessagesFactory.get(RMMessages.class);
```
Comment (risdenk) [inline gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaBaseDispatcher.java]: This has leftovers from copying elsewhere.

File: gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaDispatch.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+/**
+ * This specialized dispatch provides Livy specific features to the
+ * default dispatch.
+ */
+public class LivyHaDispatch extends LivyHaBaseDispatcher {
+  private static final String RESOURCE_ROLE = "RESOURCEMANAGER";
+  private HaProvider haProvider;  
+
+  private static final RMMessages LOG = MessagesFactory.get(RMMessages.class);
+
```
Comment (risdenk) [inline gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaDispatch.java]: This has leftovers from copying elsewhere.

File: gateway-service-definitions/src/main/resources/services/livy/0.4.0/service.xml [hunk 7]
```diff
@@ -17,9 +17,9 @@
 -->
 <service role="LIVYSERVER" name="livy" version="0.4.0">
```
Comment (risdenk) [inline gateway-service-definitions/src/main/resources/services/livy/0.4.0/service.xml]: Most likely going to want a new version of the service definitions instead of just modifying what is there. I don't think 0.4.0 of Livy had HA redirects.

File: gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaDispatch.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class  LivyHaDispatch extends DefaultHaDispatch {
+    private static final String FAILOVER_COUNTER_ATTRIBUTE = "dispatch.ha.failover.counter";
+    private static final String LOCATION = "Location";
+    private static final RMMessages LOG = MessagesFactory.get(RMMessages.class);
```
Comment (risdenk) [inline gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaDispatch.java]: Still labeled as `RMMessages`.

File: gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/i18n/RMMessages.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+package org.apache.knox.gateway.livy.i18n;
+
+import org.apache.knox.gateway.ha.dispatch.i18n.HaDispatchMessages;
+import org.apache.knox.gateway.i18n.messages.Message;
+import org.apache.knox.gateway.i18n.messages.MessageLevel;
+import org.apache.knox.gateway.i18n.messages.Messages;
+import org.apache.knox.gateway.i18n.messages.StackTrace;
+
+@Messages(logger = "org.apache.knox.gateway")
+public interface RMMessages extends HaDispatchMessages {
```
Comment (risdenk) [inline gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/i18n/RMMessages.java]: Still labeled as `RMMessages`.

File: gateway-service-definitions/src/main/resources/services/livy/0.4.0/rewrite.xml [hunk 10]
```diff
@@ -6,9 +6,7 @@
    The ASF licenses this file to You under the Apache License, Version 2.0
    (the "License"); you may not use this file except in compliance with
    the License.  You may obtain a copy of the License at
-
```
Comment (risdenk) [inline gateway-service-definitions/src/main/resources/services/livy/0.4.0/rewrite.xml]: Still changes to this file that aren't needed.

File: gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaDispatch.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+          executeRequest(outboundRequest, inboundRequest, outboundResponse);
+       } else {
+          LOG.maxFailoverAttemptsReached(getMaxFailoverAttempts(), getServiceRole());
+          if (inboundResponse != null) {
+             writeOutboundResponse(outboundRequest, inboundRequest, outboundResponse, inboundResponse);
+          } else {
+             throw new IOException(exception);
+          }
+       }
+    }
```
Comment (risdenk) [inline gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaDispatch.java]: This whole method is a copy from `DefaultHaDispatch` so don't need to copy it here?

I guess the only differnce is `getUriFromInbound` but that instead could be an overridden `getDispatchUrl` and add the custom logic there?

**Comment by risdenk** [comment]
Also the travis failures are real:

```
[INFO] --- maven-dependency-plugin:3.1.1:analyze-only (analyze-dependencies) @ gateway-service-livy ---
[WARNING] Used undeclared dependencies found:
[WARNING]    org.apache.httpcomponents:httpclient:jar:4.5.11:compile
[WARNING]    org.apache.httpcomponents:httpcore:jar:4.4.13:compile
```

File: gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaDispatch.java (hunk 8 above)
Comment (RogPodge) [reply gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaDispatch.java]: Changed

File: gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/i18n/RMMessages.java (hunk 9 above)
Comment (RogPodge) [reply gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/i18n/RMMessages.java]: changed

File: gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaDispatch.java (hunk 11 above)
Comment (RogPodge) [reply gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaDispatch.java]: refactored to inherit from DefaultHADispatch

**Comment by risdenk** [comment]
Thanks @RogPodge just noticed you pushed new changes. Github doesn't notify reviewers of new pushed commits to PRs :/ I'll take another look at this soon.
